pkg/webhook/policy.go

```diff
@@ -258,51 +258,65 @@ func (h *validationHandler) getValidationMessages(res []*rtypes.Result, req *adm
+			for _, action := range r.ScopedEnforcementActions {
+				if err := util.ValidateEnforcementAction(util.EnforcementAction(action), r.Constraint.Object); err != nil {
+					continue
+				}
+				actions = append(actions, action)
+			}
+			if len(actions) == 0 {
 				continue
 			}
-			if *logDenies {
-				h.log.WithValues(
-					logging.Process, "admission",
-					logging.Details, r.Metadata["details"],
-					logging.EventType, "violation",
-					logging.ConstraintName, r.Constraint.GetName(),
-					logging.ConstraintGroup, r.Constraint.GroupVersionKind().Group,
-					logging.ConstraintAPIVersion, r.Constraint.GroupVersionKind().Version,
-					logging.ConstraintKind, r.Constraint.GetKind(),
-					logging.ConstraintAction, r.EnforcementAction,
-					logging.ConstraintEnforcementActions, action,
-					logging.ResourceGroup, req.AdmissionRequest.Kind.Group,
-					logging.ResourceAPIVersion, req.AdmissionRequest.Kind.Version,
-					logging.ResourceKind, req.AdmissionRequest.Kind.Kind,
-					logging.ResourceNamespace, req.AdmissionRequest.Namespace,
-					logging.ResourceName, resourceName,
-					logging.RequestUsername, req.AdmissionRequest.UserInfo.Username,
-				).Info(
-					fmt.Sprintf("denied admission: %s", r.Msg))
+		default:
+			if err := util.ValidateEnforcementAction(util.EnforcementAction(r.EnforcementAction), r.Constraint.Object); err != nil {
+				continue
 			}
-			if *emitAdmissionEvents {
-				annotations := map[string]string{
-					logging.Process:                      "admission",
-					logging.EventType:                    "violation",
-					logging.ConstraintName:               r.Constraint.GetName(),
-					logging.ConstraintGroup:              r.Constraint.GroupVersionKind().Group,
-					logging.ConstraintAPIVersion:         r.Constraint.GroupVersionKind().Version,
-					logging.ConstraintKind:               r.Constraint.GetKind(),
-					logging.ConstraintAction:             r.EnforcementAction,
-					logging.ConstraintEnforcementActions: action,
-					logging.ResourceGroup:                req.AdmissionRequest.Kind.Group,
-					logging.ResourceAPIVersion:           req.AdmissionRequest.Kind.Version,
-					logging.ResourceKind:                 req.AdmissionRequest.Kind.Kind,
-					logging.ResourceNamespace:            req.AdmissionRequest.Namespace,
-					logging.ResourceName:                 resourceName,
-					logging.RequestUsername:              req.AdmissionRequest.UserInfo.Username,
-				}
+		}
+		if *logDenies {
+			h.log.WithValues(
+				logging.Process, "admission",
+				logging.Details, r.Metadata["details"],
+				logging.EventType, "violation",
+				logging.ConstraintName, r.Constraint.GetName(),
+				logging.ConstraintGroup, r.Constraint.GroupVersionKind().Group,
+				logging.ConstraintAPIVersion, r.Constraint.GroupVersionKind().Version,
+				logging.ConstraintKind, r.Constraint.GetKind(),
+				logging.ConstraintAction, r.EnforcementAction,
+				logging.ConstraintEnforcementActions, actions,
```

Comment:
does this also need `,` string delimiter

Reply:
Nope, Logging accepts `interface{}` so it should be fine. Below is the logged deny - 

```log
{"level":"info","ts":1722403563.1885743,"logger":"webhook","msg":"denied admission: you must provide labels: {\"gatekeeper\"}","hookType":"validation","process":"admission","details":{"missing_labels":["gatekeeper"]},"event_type":"violation","constraint_name":"cm-must-have-gk-scoped","constraint_group":"constraints.gatekeeper.sh","constraint_api_version":"v1beta1","constraint_kind":"K8sRequiredLabels","constraint_action":"scoped","constraint_enforcement_actions":["deny","warn"],"resource_group":"","resource_api_version":"v1","resource_kind":"ConfigMap","resource_namespace":"gatekeeper-test-playground-scoped","resource_name":"bad-cm-scoped","request_username":"kubernetes-admin"}
```